### PR TITLE
Resolve more MSVC warnings

### PR DESCRIPTION
**Pull request by olgavrou**
This should resolve the last warnings. Checked build with Visual Studio 17 and Visual Studio 19.

### CODE REVIEW

File: vowpalwabbit/array_parameters.h [hunk 1]
```diff
@@ -160,7 +160,8 @@ class sparse_parameters
     {
       for (auto& iter : _map)
       {
-        free(iter.second);
+        auto weight_ptr = iter.second;
```
Comment (jackgerrits) [inline vowpalwabbit/array_parameters.h:163]: Nit: `auto*`

File: vowpalwabbit/array_parameters.h [hunk 2]
```diff
@@ -206,7 +207,8 @@ class sparse_parameters
...
```
Comment (jackgerrits) [inline vowpalwabbit/array_parameters.h:210]: Nit: `auto*`

Comment (olgavrou) [reply vowpalwabbit/array_parameters.h:210]: is this for styling? because you don't need to declare pointers as `auto*`

Comment (jackgerrits) [reply vowpalwabbit/array_parameters.h:210]: Just a preference thing, to me it makes it clear that this is a pointer so I don't need to worry about a potential copy here.

**Pull request merged**